### PR TITLE
Fix a number of issues with locking and fsck

### PR DESCRIPTION
**Pull request by TheJare**
- fsck only scans current version of objects
- Files are made readonly after unlocking them, by path or id
- Locking paths are relative to current directory, not to repo root
- Fixed errors with unable to stat files if cwd is not the repo root
- Fixed error on OSX when filesystem path to repo contains uppercase letters. Sometimes $PWD would have them in lowercase, OSX filesystem preserves case but is case insensitive, however the various path functions in Go do not do this. Changed to ask Git itself for the repo-relative paths.

### CODE REVIEW

File: commands/command_fsck.go [hunk 1]
```diff
@@ -47,7 +47,7 @@ func fsckCommand(cmd *cobra.Command, args []string) {
 		}
 	})
 
-	if err := gitscanner.ScanRefWithDeleted(ref.Sha, nil); err != nil {
+	if err := gitscanner.ScanRef(ref.Sha, nil); err != nil {
```
Comment (ttaylorr) [inline commands/command_fsck.go]: Could you provide a little more context why this method call was changed to no longer scan through deleted objects?

My thought was that the idea of a fsck in this context was to find and remove the LFS pointers in deleted Git objects.

File: locking/locks.go [hunk 2]
```diff
@@ -154,6 +157,17 @@ func (c *Client) UnlockFileById(id string, force bool) error {
 	if err := c.cache.RemoveById(id); err != nil {
 		return fmt.Errorf("Error caching unlock information: %v", err)
 	}
+	path := unlockRes.Lock.Path
+	gitRoot, err := git.RootDir()
+	if err != nil {
+		return err
+	}
+	absPath := filepath.Join(gitRoot, path)
```
Comment (ttaylorr) [inline locking/locks.go]: The locking API guarantees that paths returned from a spec-compliant server will be absolute with the root being the root directory of the git repository.

That all being said, I think this might still be a safe (and good) thing to do, _if_ we first:

- Checked to see whether or not the `unlockRes.Lock.Path` is absolute (i.e., starts with a "/").
- If so, use that as is.
- If not, make a call to `filepath.Join` with the first argument being the `gitRoot`.

I think that this can be implemented fairly cleanly as:

```go
var lockPath = unlockRes.Lock.Path
if !path.IsAbs(lockPath) {
        gitRoot, err := git.RootDir()
        if err != nil {
                return err
        }

        lockPath = filepath.Join(gitRoot, lockPath)
}

// ...
```

File: locking/locks.go [hunk 3]
```diff
@@ -154,6 +157,17 @@ func (c *Client) UnlockFileById(id string, force bool) error {
...
+
+	// Make non-writeable if required
+	if c.SetLockableFilesReadOnly && c.IsFileLockable(path) {
```
Comment (ttaylorr) [inline locking/locks.go]: Good catch.

Can you add a test in the `test/test-unlock.sh` file to prevent this functionality from regressing in the future and add a note to the manpage to document this behavior?

File: locking/locks.go [hunk 4]
```diff
@@ -106,8 +108,13 @@ func (c *Client) LockFile(path string) (Lock, error) {
 		return Lock{}, errors.Wrap(err, "lock cache")
 	}
 
+	gitRoot, err := git.RootDir()
+	if err != nil {
+		return Lock{}, err
+	}
+	absPath := filepath.Join(gitRoot, path)
```
Comment (ttaylorr) [inline locking/locks.go]: See same note about making relative paths absolute.

File: git/git.go [hunk 5]
```diff
@@ -615,6 +615,32 @@ func GitAndRootDirs() (string, string, error) {
 	return absGitDir, absRootDir, nil
 }
 
+func PathFromRoot(pattern string) ([]string, error) {
```
Comment (ttaylorr) [inline git/git.go]: Awesome function: this makes the call site much simpler. I have two suggestions:

1. What do you think about changing the name of this function to be `PathsFromRoot` (plural) since it returns multiple paths?
2. What do you think about adding a `PathFromRoot` (singular) that returns `(string, error)` and additionally returns an error if more than one path matched?

In suggestion 2, I think the idea would be to allow `("", nil)` a valid return value and leave that check (`if len(path) == 0`) at the original location in commands/command_lock.go, but move the `if len(paths) > 1` check to this new function.

File: commands/command_fsck.go (hunk 1 above)
Comment (TheJare) [inline commands/command_fsck.go]: I am not clear on the goal of fsck, but when I ran it on my repo it complained about missing objects that were old revisions of lfs-stored files. It then tried to move the offending objects' files to the "bad" folder and, since the files of course did not exist, it complained again.

I found this looked like a reasonable fix, and then saw #953 which seems to agree. I should probably have left this in another pull request, but I'm not very well versed in proper pull requesting. :)

File: git/git.go (hunk 5 above)
Comment (TheJare) [inline git/git.go]: I considered making the function just return one path or error if there are more, but it seems to me the evolution of the lock commands will/should include supporting multiple files at a time (e.g. `git lfs lock a.jpg b.jpg`, or `git lfs lock *.zip` so returning multiple paths directly and letting be up to the caller to decide about supporting multiple files or just one.

On the other hand, the function as it stands will match many more files than one would initially guess, so if one passes "a.zip" as argument I think the function will return "a.zip" files in folder and subfolders. This will only cause more trouble if git-lfs starts supporting wildcard arguments ("*" would lock the entire repo!). So this needs more design attention, but right now seems like an improvement over the previous functionality and I didn't want to complicate it with more API points.

File: locking/locks.go (hunk 3 above)
Comment (TheJare) [inline locking/locks.go]: Not sure if this comment refers to initializing the Config.SetLockableFilesReadOnly to true by default? If so, I believe the docs already say that's how it works. I assumed the config variable (and env var) were added as part of a future method for configuring this behaviour. I noticed other places that do not test the flag and were already setting the readonly flag (despite the config defaulting to false), so there's more fixes to be done.

I'll look at adding the test.

File: locking/locks.go (hunk 2 above)
Comment (TheJare) [inline locking/locks.go]: At the time I did this I was testing with my own fork of lfs-test-server since there were no other servers (not even github's own) that were supporting locking. I noticed yesterday as I was working on this that Github locks were starting to work (and could see how they were adding/fixing stuff on the fly as I tested my git-lfs work on Github as well as my fork). I'll review this because my fork may not be adhering to that aspect of the API.

File: commands/command_fsck.go (hunk 1 above)
Comment (ttaylorr) [inline commands/command_fsck.go]: Ah, thanks for linking that -- I forgot about that issue. I agree that the fsck changes could come in a separate pull request, since it seems there are more options re: different ways that that could work.

How do you feel about backing out of this change for now and moving it into a different PR?

File: locking/locks.go (hunk 2 above)
Comment (ttaylorr) [inline locking/locks.go]: > I'll review this because my fork may not be adhering to that aspect of the API.

Sounds good. I think my preference would be to back out of this change as well, but it may be good to keep anyway, so that LFS can better handle servers that implement the locking API incorrectly.

What are your thoughts?

File: locking/locks.go (hunk 3 above)
Comment (ttaylorr) [inline locking/locks.go]: > I noticed other places that do not test the flag and were already setting the readonly flag (despite the config defaulting to false), so there's more fixes to be done.
> 
> I'll look at adding the test.

That sounds good, this was my main area of concern.

File: git/git.go (hunk 5 above)
Comment (ttaylorr) [inline git/git.go]: > I considered making the function just return one path or error if there are more, but it seems to me the evolution of the lock commands will/should include supporting multiple files at a time

I think that it should be the caller's (in this case, the `commands` package, and particularly, `commands/commands_lock.go`) to break up multiple arguments into multiple calls. I think this could look something like:

```go
for _, lpath := range args {
        path, err := git.PathFromRoot(lpath)
        // ...
}
```

That being said, I think I still prefer my original suggestion of having both:

- `PathFromRoot(pattern string) (path string, err error)`
- `PathsFromRoot(pattern string) (paths []string, err error)`

Would you be opposed to going forward with this?
